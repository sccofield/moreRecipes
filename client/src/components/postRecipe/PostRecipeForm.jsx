import React from 'react';
import Dropzone from 'react-dropzone';
import PropTypes from 'prop-types';


const PostRecipeForm = props => (
  <main className="container">

    <div className="row justify-content-md-center">
      <div className="col-md-8">

        <h1>Add New Recipe</h1>
        {props.state.errors &&
          <ul className="errorMessage">{
            props.state.errors.map(error => (
              <li key={Math.floor(Math.random() * 1000)
              }
              >{error}
              </li>))}
          </ul>
        }

        <form onSubmit={props.onSubmit}>
          <div className="form-group">
            <label htmlFor="recipeTitle">Recipe title</label>
            <input
              type="text"
              className="form-control"
              name="title"
              id="recipeTitle"
              placeholder="Enter recipe title"
              value={props.state.title}
              onChange={props.onChange}
              required
            />
          </div>
          <div className="form-group">
            <label htmlFor="ingredients">Recipe ingredients</label>

            <textarea
              className="form-control"
              name="ingredients"
              id="ingredients"
              placeholder="Enter recipe ingredients seperated by commas"
              rows="5"
              value={props.state.ingredients}
              onChange={props.onChange}
              required
            />
            <small id="emailHelp" className="form-text text-muted">
              seperate each ingredients with a comma (,)
            </small>

          </div>

          <div className="form-group">
            <label htmlFor="post">Recipe description</label>
            <textarea
              className="form-control"
              name="description"
              id="description"
              rows="5"
              placeholder="Enter recipe description"
              value={props.state.description}
              onChange={props.onChange}
              required
            />
          </div>

          <div className="form-group">
            <label htmlFor="imgurl">Recipe image</label>
            <Dropzone
              onDrop={props.handleDrop}
              multiple={false}
            >
              <p className="text-center pt-4">
                Click to upload image
              </p>
              { props.imagePreview &&

              <div className="form-group">
                <img
                  src={props.imagePreview}
                  alt=""
                  className="img img-responsive img-thumbnail imagePreview"
                />
              </div>
              }
            </Dropzone>


          </div>


          <button type="submit" className="btn btn-primary">Submit</button>
        </form>

      </div>
    </div>
  </main>

);

PostRecipeForm.propTypes = {
  errors: PropTypes.arrayOf(PropTypes.string).isRequired,
  handleDrop: PropTypes.func.isRequired,
  imagePreview: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  state: PropTypes.shape({
    description: PropTypes.string,
    errors: PropTypes.array,
    image: PropTypes.string,
    ingredients: PropTypes.string,
    title: PropTypes.string
  }).isRequired,
};

PostRecipeForm.defaultProps = {
  imagePreview: null
};

export default PostRecipeForm;
