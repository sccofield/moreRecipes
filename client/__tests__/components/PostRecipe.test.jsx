import React from 'react';
import toJSON from 'enzyme-to-json';
import sinon from 'sinon';
import moxios from 'moxios';
import { PostRecipe, mapDispatchToProps, mapStateToProps }
  from '../../src/components/postRecipe/PostRecipe';
import PostRecipeForm from '../../src/components/postRecipe/PostRecipeForm';

describe('Post recipe controller', () => {
  beforeEach(() => {
    moxios.install();
  });
  afterEach(() => {
    moxios.uninstall();
  });

  const props = {
    postRecipeActionCreator: jest.fn(() => Promise.resolve()),
    clearNewRecipe: jest.fn(),
    history: {
      push: jest.fn()
    },

  };


  it('should render', () => {
    const wrapper = shallow(<PostRecipe {...props} />);
    expect(wrapper.find('form').exists).toBeTruthy();
    expect(wrapper.find(PostRecipeForm).exists).toBeTruthy();
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it(
    'sets the state when the value of the input changes',
    () => {
      const spy = sinon.spy(PostRecipe.prototype, 'onChange');
      const wrapper = shallow(<PostRecipe {...props} />);
      const event = {
        target: {
          value: 'recipe title',
          name: 'title'
        }
      };
      wrapper.instance().onChange(event);
      expect(wrapper.state('title')).toBe('recipe title');

      expect(spy.called).toBeTruthy();
    }
  );
});

describe('conatiner functions', () => {
  test('mapStateToProps', () => {
    expect(mapStateToProps({
      recipe: { }
    })).toHaveProperty('newRecipe');
  });

  test('mapDispatchToProps', () => {
    const dispatch = jest.fn();
    expect(mapDispatchToProps(dispatch))
      .toHaveProperty('postRecipeActionCreator');

    const { postRecipeActionCreator } = mapDispatchToProps(dispatch);
    postRecipeActionCreator();
    expect(dispatch).toHaveBeenCalled();
  });
});

