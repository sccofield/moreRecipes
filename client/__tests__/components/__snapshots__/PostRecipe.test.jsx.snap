// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post recipe controller should render 1`] = `
<div
  className="main"
>
  <Connect(PageHeader) />
  <div
    className="container"
  >
    <div
      className="row content"
    >
      <div
        className="col-lg-10"
      >
        <PostRecipeForm
          className="mb-4"
          errorMessage={null}
          errors={Array []}
          handleDrop={[Function]}
          imagePreview={null}
          onChange={[Function]}
          onSubmit={[Function]}
          state={
            Object {
              "description": "",
              "errors": Array [],
              "image": null,
              "ingredients": "",
              "title": "",
            }
          }
        />
      </div>
    </div>
  </div>
  <Footer />
</div>
`;
